File: Problem35_3Sum.java
// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> triplets = new ArrayList<>();
        if(nums == null || nums.length == 0)
            return triplets;
        Arrays.sort(nums);
        for(int i = 0; i < nums.length; i++) {
            if((i == 0) || (i > 0 && nums[i] != nums[i - 1])) {
                int left = i + 1;
                int right = nums.length - 1;
                while(left < right) {
                    int sum = nums[i] + nums[left] + nums[right];
                    if(sum == 0) {
                        triplets.add(Arrays.asList(nums[i], nums[left], nums[right]));
                        while(left < right && nums[left] == nums[left + 1])
                            left++;
                        while(left < right && nums[right] == nums[right - 1])
                            right--;
                        left++;
                        right--;
                    } else if(sum < 0)
                        left++;
                    else
                        right--;
                }
            }
        }
        return triplets;
    }
}

File: Problem 35 3Sum.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n^2)
# Space Complexity: O(n) bcz if sorting
# Logic: Extend the approach to calculate 2 sum, sort the array keep i to the left most & for every i calculate sum for
# every j & k. if sum is 0 add in output else keep on checking. To avoid duplicates in output skip repeating elements.

class Solution:
    def threeSum(self, nums):
        self.output = []
        nums.sort()
        for i in range(len(nums)-2):
            if nums[i] > 0:
                break
            if i == 0 or nums[i] != nums[i-1]:
                self.twoSum(nums, i)
        return self.output

    def twoSum(self, nums, i):
        j = i + 1
        k = len(nums) - 1
        while j < k:
            sum = nums[i] + nums[j] + nums[k]
            if sum == 0:
                self.output.append([nums[i], nums[j], nums[k]])
                j += 1
                k -= 1
                while j < k and nums[j] == nums[j-1]:
                    j += 1
            elif sum < 0:
                j += 1
            else:
                k -= 1